# Right Most set bit 

In [13]:
def rsb(num):
    ans=num & -num
    return int(bin(ans).replace('0b',''))

In [17]:
rsb(122)

10

# kernighan's Algorithm 
total on (1) bits

In [18]:
def kernighan(num):
    counter=0
    while num!=0:
        rsbm=num & -num
        num-=rsbm
        counter+=1
    print('bits on:',counter)

In [20]:
kernighan(13)

bits on: 3


# Josephus Problem

In [40]:
def powerof2(n):
    i=1
    while i*2<=n:
        i=i*2
    return i
def josephus(n):
    pow2=powerof2(n)
    l=n-pow2
    return 2*l+1

In [45]:
josephus(12)

9

# N bits grey code 

In [23]:
def greycode(n):
    if n==1:
        return ['0','1']
    ros=greycode(n-1)
    result=[]
    for i in range(0,len(ros)):
        result.append('0'+ros[i])
    for i in range(len(ros)-1,-1,-1):
        result.append('1'+ros[i])
    return result

In [24]:
greycode(3)

['000', '001', '011', '010', '110', '111', '101', '100']

# Minimum no of Software developers 
given n skills and k software developers with thier skills tell how many developers need to have all skills

In [86]:
skills=['java','python','nodejs','cpp','marketing']
skillmap={skills[i]:i for i in range(0,len(skills))}
person=[i for i in range(4)]
personskill={0:['java','python'],1:['cpp'],2:['marketing','nodejs','cpp'],3:['java','nodejs'],4:['marketing']}
for i in person:
    for skill in personskill[i]:
        person[i]=person[i]| (1<<skillmap[skill])
sol=[]
def solution(people,nskills,cp,onsol,smask):
    global sol
    if cp==len(people):
        if smask == ((1<<nskills)-1):
            if len(sol)==0 or len(onsol)<len(sol):
                sol=onsol
        return
    solution(people,nskills,cp+1,onsol,smask)
    
    onsol.append(cp)
    solution(people,nskills,cp+1,onsol,(smask|people[cp]))
    onsol.pop()
solution(person,5,0,[],0)

# All repeating except two 

In [14]:
def repeating(arr):
    xxor=0
    for val in arr:
        xxor=xxor^val
    
    rmsb=xxor&-xxor
    x=0
    y=0
    for val in arr:
        if val & rmsb==0:
            x=x^val
        else:
            y=y^val
    if x<y:
        print(x)
        print(y)
    else:
        print(y)
        print(x)

In [17]:
repeating([1,1,2,2,3,3,9,4,4,5,5,6])

6
9


# Find duplicate No. and Missing No from 1 to n 

In [26]:
def dupandmiss(arr):
    xor=0
    for val in arr:
        xor^=val
    for i in range(1,len(arr)+1):
        xor^=val
        
    rmsb=xor&-xor
    x=0
    y=0
    for val in arr:
        if rmsb&val==0:
            x^=val
        else:
            y^=val
    
    for i in range(1,len(arr)+1):
        if rmsb&i==0:
            x^=i
        else:
            y^=i
    print(x,y)
    for val in arr:
        if val==x:
            print('missing no->',y)
            print('duplicate no->',x)
            break
        elif val==y:
            print('missing no->',x)
            print('duplicate no->',y)
            break
dupandmiss([1,2,2,4,5])

2 3
missing no-> 3
duplicate no-> 2


# find element that appear once 

In [27]:
def unique(arr):
    xor=0
    for val in arr:
        xor^=val
    return xor

In [28]:
unique([1,1,2,2,3,4,4,5,5])

3

# Xor of sum of all pairs 

In [29]:
def xorsum(arr):
    xor=0
    for val in arr:
        xor=xor^2*val
    return xor
xorsum([1,2,3,4])

8

# Reduce N to 1 
minimum operations to reduce n to 1 

In [35]:
def reduce1(num):
    res=0
    while num!=1:
        if num%2==0:
            num=num//2
        elif num==3:
            num=num-1
        elif num%4==1:
            num=num-1
        elif num%4==3:
            num=num+1
        res+=1
    return res
reduce1(8)

3

# find no of triplets 

# Print Abbrevation 

In [43]:
def printabbrevation(string):
    for i in range(0,1<<len(string)):
        newstr=''
        count=0
        for j in range(0,len(string)):
            ch=string[j]
            b=len(string)-1-j
            if i&(1<<b)==0:
                if count==0:
                    newstr+=ch
                else:
                    newstr+=str(count)
                    newstr+=ch
                    count=0
            else:
                count+=1
        if count>0:
            newstr+=str(count)
        print(newstr,end='')
        print('')

In [47]:
printabbrevation('pep')

pep
pe1
p1p
p2
1ep
1e1
2p
3


# Sudoku using bit manipulation 

In [56]:
row=[0 for x in range(9)]
col=[0 for x in range(9)]
arr=[[0 for x in range(9)] for x in range(9)]
grid=[[0 for x in range(3)] for x in range(3)]
for i in range(9):
    for j in range(9):
        digit=int(input())
        arr[i][j]=digit
        row[i]|=(1<<digit)
        col[j]|=(1<<digit)
        grid[i//3][j//3]|=(1<<digit)

0
0
8
0
0
0
3
0
0
1
0
0
0
0
3
0
0
2
3
0
0
9
0
0
7
8
0
0
0
4
5
9
0
2
6
0
0
0
0
1
3
2
9
0
4
0
1
0
7
6
0
0
0
0
9
2
6
0
0
0
5
4
0
0
0
0
0
5
0
0
0
0
0
5
1
0
0
0
0
0
0


In [61]:
row=[0 for x in range(9)]
col=[0 for x in range(9)]
grid=[[0 for x in range(3)] for x in range(3)]
for i in range(0,len(arr)):
    for j in range(0,len(arr[0])):
        digit=arr[i][j]
        row[i]|=(1<<digit)
        col[j]|=(1<<digit)
        grid[i//3][j//3]|=(1<<digit)

In [68]:
def solvesudoku(arr,row,col,grid,i,j):
    if i==len(arr):
        for i in range(0,len(arr)):
            for j in range(0,len(arr[0])):
                print(arr[i][j],end=' ')
            print('')
        return
    
    ni=0
    nj=0
    if j==len(arr[0])-1:
        ni=i+1
        nj=0
    else:
        ni=i
        nj=j+1
    if arr[i][j]>0:
        solvesudoku(arr,row,col,grid,ni,nj)
    else:
        for num in range(1,10):
            if (row[i]& 1<<num)==0 and (col[j]& 1<<num)==0 and (grid[i//3][j//3]&1<<num)==0:
                arr[i][j]=num
                row[i]^=(1<<num)
                col[j]^=(1<<num)
                grid[i//3][j//3]^=(1<<num)
                solvesudoku(arr,row,col,grid,ni,nj)
                arr[i][j]=0
                row[i]^=(1<<num)
                col[j]^=(1<<num)
                grid[i//3][j//3]^=(1<<num)

In [69]:
solvesudoku(arr,row,col,grid,0,0)

5 6 8 4 2 7 3 1 9 
1 9 7 6 8 3 4 5 2 
3 4 2 9 1 5 7 8 6 
7 3 4 5 9 8 2 6 1 
6 8 5 1 3 2 9 7 4 
2 1 9 7 6 4 8 3 5 
9 2 6 8 7 1 5 4 3 
4 7 3 2 5 6 1 9 8 
8 5 1 3 4 9 6 2 7 


In [70]:
0b0

0

# UTF Ecoding 
1byte start from 10------
2byte start from 110----- 10------
3byte start from 1110--- 10----- 10-----
4byte start from 11110-- 10---- 10--- 10----

In [77]:
def checkutf(arr):
    rbyte=0
    for val in arr:
        if rbyte==0:
            if val>>7==0b0:
                rbyte=0
            elif val>>5==0b110:
                rbyte=1
            elif val>>4==0b1110:
                rbyte=2
            elif val>>3==0b11110:
                rbyte=3
            
        else:
            if val>>6==0b10:
                rbytes-=1
            else:
                return False
        return rbyte==0
checkutf([-2,-7,-5])

True

# Flip bits to convert A and B count

In [80]:
def flipbitscount(a,b):
    xorab=a^b
    #apply kernighan algo to count bits
    counter=0
    while xorab!=0:
        rsbm=xorab & -xorab
        xorab-=rsbm
        counter+=1
    return counter
flipbitscount(4,15)

3

# copy set bits in range 

In [84]:
def copyset(a,b,left,right):
    m1=(1<<(right-left+1))
    m1=m1-1
    m1=m1<<(left-1)
    m1=m1&a
    b=b|m1
    print(b)
copyset(7,5,2,4)

7


# Power of 2 

In [93]:
#1<<x =2**x
#multiply n<<1 and divide n>>1
def power2(n):
    val=n&(n-1)
    return val==0
power2(16)    

True

# Solve 7n by 8 

In [96]:
def solve7n(n):
    print(((n<<3)-n)>>3)
    return
solve7n(5)

4


# swap all odd & even bits

In [107]:
def swapevenoddbits(num):
    evmask=0xAAAAAAAA
    odmask=0x55555555
    odd=num&odmask
    even=num&evmask
    odd<<=1
    even>>=1
    num=odd|even
    print(num)

In [109]:
swapevenoddbits(5)

10


# Sum of bit difference among all pairs 

In [110]:
def sumbitdiff(arr):
    res=0
    for i in range(0,32):
        counton=0
        for val in arr:
            if val&(i<<1)!=0:
                counton+=1
        countof=len(arr)-counton
        diff=countof * counton *2
        res+=diff
    return res

In [111]:
sumbitdiff([12,34,56,7])

108

# Count Bits on in first N Natural Number

In [118]:
def largestpowerofx(n):
    x=0
    while(1<<x)<=n:
        x+=1
    return x-1
def countonbit(n):
    if n==0:
        return 0
    x=largestpowerofx(n)
    bitsupto=x*(1<<(x-1))
    msb=n-(1<<x)+1
    ans=bitsupto+msb+countonbit(n-(1<<x))
    return ans

In [120]:
countonbit(18)

37

# Reverse Bits  

In [3]:
def reversebits(n):
    j=0
    rev=0
    flag=False
    for i in range(31,-1,-1):
        mask=(1<<i)
        if flag==True:
            if n&mask!=0:
                print("1",end=' ')
                rev|=(1<<j)
            else:
                print("0",end=' ')
            j+=1
        else:
            if n&mask!=0:
                flag=True
                print("1",end=' ')
                smask=(1<<j)
                rev|=smask
                j+=1
    print('reverse bit:',rev)

In [4]:
reversebits(11)

1 0 1 1 reverse bit: 13


# Min Xor pair 

In [15]:
def minxorpair(arr):
    arr.sort()
    _min=100000
    ls=[]
    for i in range(1,len(arr)):
        pair=arr[i-1]^arr[i]
        
        if pair<_min:
            _min=pair
            ls=[]
            ls.append(['min xor pair:'+str(arr[i-1])+' and '+str(arr[i])])
        elif pair==_min:
            ls.append(['min xor pair:'+str(arr[i-1])+' and '+str(arr[i])])
        
    for x in ls:
        print(x)

In [16]:
minxorpair([3,6,1,8,5])

['min xor pair:1 and 3']
